### PR TITLE
Cirrus Terminal support

### PR DESCRIPTION
**Pull request by edigaryev**
Now tasks can be re-run with Cirrus Terminal access:

cirrus-terminal

The terminal becomes visible once the agent publishes terminal credentials and they become available through GraphQL.

Resolves https://github.com/cirruslabs/cirrus-ci-docs/issues/432.

### CODE REVIEW

File: src/components/cirrus-terminal/CirrusTerminal.ts [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    grpc.setDefaultTransport(grpc.WebsocketTransport());
+
+    // Request a new channel on the terminal server
+    const frontendService = new GuestServiceClient(connectTo);
+    this.terminalChannel = frontendService.terminalChannel();
+
+    // Do I/O
+    this.terminalChannel.on('data', message => {
+      if (this.state !== CirrusTerminalState.Connected) {
+        this.state = CirrusTerminalState.Connecting;
```
[the remark this answers is not in the document]
Comment (edigaryev) [reply src/components/cirrus-terminal/CirrusTerminal.ts:49]: > Do you want to make it Connected if it hasn't yet?

This makes sure we drop spurious I/O until the handshake below is done.

Actually it works just fine without the `state` variable, it's just that in the next iteration we can display this state to the user to help with debugging potential issues.

[the remark this answers is not in the document]
Comment (fkorotkov) [reply src/components/cirrus-terminal/CirrusTerminal.ts:49]: How does it make it drop I/O?

[the remark this answers is not in the document]
Comment (edigaryev) [reply src/components/cirrus-terminal/CirrusTerminal.ts:49]: Oh, right, added a missing `return` in 501aa11.

**Review by fkorotkov — APPROVED** [review]

**Comment by sio** [comment]
Thanks for adding this feature!

I've just tried it and it works, but the terminal session is stopped abruptly when all other tasks have ran to the end. It is not very useful, because there is no way to debug a failed task if terminal is terminated immediately after CI failure. Maybe there should be some delay in "Wait for terminal" that checks if interactive session is still in use?

**Comment by fkorotkov** [comment]
@sio that should be fixed now

**Pull request merged**